### PR TITLE
[f] Add more mock assertions

### PR DESCRIPTION
**Pull request by catriuspham**
Issue: https://github.com/EastAgile/robber.py/issues/26

We will add these expectations to our code base:

- [x] expect(mock).to.be.called_once()
- [x] expect(mock).to.be.called_with(params)
- [x] expect(mock).to.be.called_once_with(params)
- [x] expect(mock).to.have.been.ever_called_with(params) (assert_any_call)

### CODE REVIEW

**Comment by catriuspham** [comment]
@hieueastagile @NhanHo Please take a look at this PR when you have time

**Comment by hieueastagile** [comment]
@catriuspham please set the base of the branch to the `negating matcher` branch, it would be easier to review.

File: README.md [hunk 1]
```diff
@@ -235,15 +235,48 @@ expect(any_callable).to.throw(Exception)
...
+expect(mock).to.be.called_once_with(*args, **kwargs)
+```
+
+#### ever_called_with
+
+Asserts that a mock has ever been called with params. 
+The call is not necessary to be to latest one (the same as assert.any_call).
+
+```python
+expect(mock).to.have.ever.been.ever_called_with(*args, **kwargs)
```
Comment (tklarryonline) [inline README.md]: This writing is somewhat redundant. Could we rewrite it with "have been ever called with"? Or, try to support the case of `have.ever.been.called_with` alone?

File: robber/helper.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        return expected_params_str
+
+    @staticmethod
+    def build_called_params_string(call_args):
+        if str(call_args) == 'call()':
+            called_params_str = 'no arguments'
+        else:
+            called_params_str = str(call_args)[5:][:-1]
+
+        return called_params_str
```
Comment (tklarryonline) [inline robber/helper.py]: This should just return early:

```py
if str(call_args) == 'call()':
    return 'no arguments'

return str(call_args)[5:][:-1]
```

Comment (tklarryonline) [reply robber/helper.py]: Besides, 'call()', 5 and -1 look magic to me.

File: robber/helper.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class Helper:
+    @staticmethod
+    def build_expected_params_string(expected, args, kwargs):
+        if not expected:
+            expected_params_str = 'no arguments'
```
Comment (tklarryonline) [inline robber/helper.py]: This should be returned soon to omit the `else` part and make it easier to read.

File: README.md (hunk 1 above)
Comment (tklarryonline) [reply README.md]: I saw that you indeed did write `have.been.ever_called_with` in the class's documentation. So I guess this is a typo maybe?

Comment (catriuspham) [reply README.md]: It is a typo, thanks for pointing it out hehe.

File: robber/helper.py (hunk 2 above)
Comment (catriuspham) [reply robber/helper.py]: Yeah, I don't like these code also. But that's the best I can do. Do you have any suggestion? 
To be more specific, `if str(call_args) == 'call()':` is to check if there is any param in the call_args, a better approach is to check if both `call_args[0]` and `call_args[1]` is empty, but that's just too long. 
`str(call_args)[5:][:-1]` is just some basic substring getting, I just cannot find out another better way.

Comment (hieueastagile) [reply robber/helper.py]: Hm, why not `str.replace('call(', '').replace(')', '')`? And if you worry that `call(` and `)` appear in the params, just use regex for it.

Comment (tklarryonline) [reply robber/helper.py]: +1 for using Regex. It's hard for newcomers but easier to read later on.

Comment (tklarryonline) [reply robber/helper.py]: Something like extracting the stuff between `()` then if it's empty just return 'no arguments' 😀

**Comment by catriuspham** [comment]
@oyster Please help merge this PR and close this issue: https://github.com/EastAgile/robber.py/issues/26 thanks

**Pull request merged**